File: studyq-backend/routes/upload_routes.py
from flask import Blueprint, request, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity
from werkzeug.utils import secure_filename

from models import db
from models.document import Document
from services.pdf_service import extract_text_from_pdf
from services.ai_service import generate_notes
from utils.helpers import success_response, error_response, allowed_file

upload_bp = Blueprint("upload", __name__)


@upload_bp.route("/upload", methods=["POST"])
@jwt_required()
def upload_document():
    user_id = int(get_jwt_identity())

    if "file" not in request.files:
        return error_response("No file part in the request.", 400)

    file = request.files["file"]
    if file.filename == "":
        return error_response("No file selected.", 400)

    allowed = current_app.config["ALLOWED_EXTENSIONS"]
    if not allowed_file(file.filename, allowed):
        return error_response("Only PDF files are allowed.", 400)

    filename = secure_filename(file.filename)
    pdf_bytes = file.read()

    # Extract text
    try:
        extracted_text = extract_text_from_pdf(pdf_bytes)
    except Exception as e:
        return error_response(f"Failed to process PDF: {str(e)}", 422)

    # Require at least 50 characters to ensure it's not just a scanned image or empty
    if not extracted_text or len(extracted_text.strip()) < 50:
        return error_response(
            "The uploaded PDF does not contain enough readable text. "
            "Please ensure it is a text-based document and not just scanned images.", 
            422
        )

    # Generate notes asynchronously in background (best-effort)
    notes = None
    try:
        notes = generate_notes(extracted_text)
    except Exception:
        pass  # Notes are optional; won't block upload

    doc = Document(
        user_id=user_id,
        file_name=filename,
        extracted_text=extracted_text,
        notes=notes,
    )
    db.session.add(doc)
    db.session.commit()

    return success_response(doc.to_dict(), "Document uploaded successfully.", 201)


@upload_bp.route("/documents", methods=["GET"])
@jwt_required()
def list_documents():
    user_id = int(get_jwt_identity())
    docs = (
        Document.query.filter_by(user_id=user_id)
        .order_by(Document.uploaded_at.desc())
        .all()
    )
    return success_response([d.to_dict() for d in docs])


@upload_bp.route("/documents/<int:doc_id>", methods=["GET"])
@jwt_required()
def get_document(doc_id):
    user_id = int(get_jwt_identity())
    doc = Document.query.filter_by(id=doc_id, user_id=user_id).first()
    if not doc:
        return error_response("Document not found.", 404)
    return success_response(doc.to_dict(include_text=True))

File: studyq-backend/routes/uploadRoutes.js
const express = require('express');
const multer = require('multer');
const Document = require('../models/Document');
const { extractTextFromPdf } = require('../services/pdfService');
const { generateNotes } = require('../services/aiService');
const { successResponse, errorResponse } = require('../utils/helpers');
const authMiddleware = require('../middleware/auth');

const router = express.Router();

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: parseInt(process.env.MAX_CONTENT_LENGTH_MB || 50) * 1024 * 1024 }
});

router.post('/upload', authMiddleware, upload.single('file'), async (req, res) => {
  try {
    const file = req.file;
    if (!file) {
      return errorResponse(res, "No file selected.", 400);
    }

    if (!file.originalname.toLowerCase().endsWith('.pdf')) {
      return errorResponse(res, "Only PDF files are allowed.", 400);
    }

    let extractedText;
    try {
      extractedText = await extractTextFromPdf(file.buffer);
    } catch (err) {
      return errorResponse(res, `Failed to process PDF: ${err.message}`, 422);
    }

    if (!extractedText || extractedText.trim().length < 50) {
      return errorResponse(
        res,
        "The uploaded PDF does not contain enough readable text. Please ensure it is a text-based document and not just scanned images.",
        422
      );
    }

    let notes = null;
    try {
      notes = await generateNotes(extractedText);
    } catch (err) {
      // Notes are optional; won't block upload
    }

    const doc = new Document({
      user_id: req.user.id,
      file_name: file.originalname,
      extracted_text: extractedText,
      notes: notes
    });
    await doc.save();

    return successResponse(res, doc.toDict(false, 0), "Document uploaded successfully.", 201);
  } catch (err) {
    console.error(err);
    return errorResponse(res, "Server error during upload.", 500);
  }
});

router.get('/documents', authMiddleware, async (req, res) => {
  try {
    const docs = await Document.find({ user_id: req.user.id }).sort({ uploaded_at: -1 }).populate('quizzes');
    const responseData = docs.map(d => d.toDict(false, d.quizzes ? d.quizzes.length : 0));
    return successResponse(res, responseData);
  } catch (err) {
    console.error(err);
    return errorResponse(res, "Server error fetching documents.", 500);
  }
});

router.get('/documents/:doc_id', authMiddleware, async (req, res) => {
  try {
    const doc = await Document.findOne({ _id: req.params.doc_id, user_id: req.user.id }).populate('quizzes');
    if (!doc) {
      return errorResponse(res, "Document not found.", 404);
    }
    return successResponse(res, doc.toDict(true, doc.quizzes ? doc.quizzes.length : 0));
  } catch (err) {
    console.error(err);
    return errorResponse(res, "Server error fetching document.", 500);
  }
});

module.exports = router;
